src/group_theory/subgroup.lean

```diff
@@ -513,6 +517,28 @@ lemma mem_Sup_of_mem {S : set (subgroup G)} {s : subgroup G}
   (hs : s ∈ S) : ∀ {x : G}, x ∈ s → x ∈ Sup S :=
 show s ≤ Sup S, from le_Sup hs
 
+/--
+Unfortunately the `mem_sup` down below only works with `comm_(add)group G`,
+so for now we must bear with the confusing names.
+-/
+@[to_additive "Unfortunately the `mem_sup` down below only works with `comm_addgroup G`,
+so for now we must bear with the confusing names."]
+lemma mem_sup_iff {H K : subgroup G} {g : G} (h : ↑(H ⊔ K) = (H : set G) * K) :
+  g ∈ H ⊔ K ↔ ∃ x y, x ∈ H ∧ y ∈ K ∧ x * y = g :=
+set.ext_iff.1 h g
+
+/--
+Unfortunately the `mem_sup'` down below only works with `comm_group G`,
+so for now we must bear with the confusing names.
+-/
+@[to_additive "Unfortunately the `mem_sup'` down below only works with `comm_addgroup G`,
+so for now we must bear with the confusing names."]
+lemma mem_sup_iff' {H K : subgroup G} {g : G} (h : ↑(H ⊔ K) = (H : set G) * K) :
+  g ∈ H ⊔ K ↔ ∃ (x : H) (y : K), (x * y : G) = g :=
+(mem_sup_iff h).trans
+  ⟨λ ⟨a, b, ha, hb, h⟩, ⟨⟨a, ha⟩, ⟨b, hb⟩, h⟩,
+  λ ⟨⟨a, ha⟩, ⟨b, hb⟩, h⟩, ⟨a, b, ha, hb, h⟩⟩
```

Comment:
These lemma look quite strange to me, especially the first one which is really just a single call to `set.ext_iff`

Reply:
I mostly use it to fight against lots of defeq abuse. 

e.g. here:
```lean
noncomputable def Zassenhaus_fun_aux {A' A : subgroup G} (B' B : subgroup G) (hA : A' ≤ A)
  [hAN : (A'.subgroup_of A).normal] (x : A' ⊔ A ⊓ B) : Zassenhaus_quot A' A B' B :=
quotient.mk' ⟨_, ((mem_sup_iff (Zassenhaus_aux B hA)).mp x.2).some_spec.some_spec.2.1⟩
```
The `_` hole is something like `((mem_sup_iff (Zassenhaus_aux B hA)).mp x.2).some_spec.some` and I think it's easier with `mem_sup_iff` adding in lots of extra type info the elaborator can work with.